### ISSUE TITLE
Is this operation equivalent?

### ISSUE THREAD

**Issue by Lokiscripter** [yongliang-wu/DFT#5]
When use vocab_parallel_cross_entropy as the loss_fct, is this operation equivalent?

losses = tensor_parallel.vocab_parallel_cross_entropy(shift_logits, shift_labels)

# from your implement
                probs = torch.softmax(shift_logits, dim=-1)
                prob_coefficients = probs.gather(1, shift_labels.unsqueeze(-1)).squeeze(-1)

# another implement

                prob_coefficients = torch.exp(-losses)

**Comment by yongliang-wu** [comment]
Yes, I think they should be equivalent under normal settings.

**Comment by yongliang-wu** [comment]
I'm not sure about these... But the DFT will have a lower loss than SFT.

**Comment by Lokiscripter** [comment]
> I'm not sure about these... But the DFT will have a lower loss than SFT.

Do you know roughly by how many orders of magnitude it is smaller?

**Comment by 1787648106** [comment]
> > Yes, I think they should be equivalent under normal settings.
> 
> My labels contain some negative elements, so my implement is below:
> 
> # my loss shape is [B,S]
> 
>             flat_labels = labels.view(-1).contiguous() 
>             if flat_labels.dtype != torch.long:
>                 flat_labels = flat_labels.long()
>             flat_losses = losses.view(-1)
>             prob_coefficients = torch.exp(-flat_losses)
>             if (flat_labels < 0).any():
>                 valid_mask = (flat_labels >= 0).float()
>                 prob_coefficients = prob_coefficients * valid_mask.to(prob_coefficients.device)
>             weighted_losses = flat_losses * prob_coefficients.detach()
>             weighted_losses = weighted_losses.view_as(losses)
>             losses = weighted_losses
> but the loss is very small like 0.0001, I'm not sure if it's normal

hello, 为啥不直接 flat_losses * torch.exp(-flat_losses).detach()呀，y对应位置的prob已经通过右侧的torch.exp算出来了，直接按位相乘再求和或求均值就行了吧

**Comment by yongliang-wu** [comment]
> > I'm not sure about these... But the DFT will have a lower loss than SFT.
> 
> Do you know roughly by how many orders of magnitude it is smaller?

From my observations, about 10-20 times smaller than SFT loss.

**Comment by Lokiscripter** [comment]
> > > Yes, I think they should be equivalent under normal settings.
> > 
> > 
> > My labels contain some negative elements, so my implement is below:
> > # my loss shape is [B,S]
> > ```
> >         flat_labels = labels.view(-1).contiguous() 
> >         if flat_labels.dtype != torch.long:
> >             flat_labels = flat_labels.long()
> >         flat_losses = losses.view(-1)
> >         prob_coefficients = torch.exp(-flat_losses)
> >         if (flat_labels < 0).any():
> >             valid_mask = (flat_labels >= 0).float()
> >             prob_coefficients = prob_coefficients * valid_mask.to(prob_coefficients.device)
> >         weighted_losses = flat_losses * prob_coefficients.detach()
> >         weighted_losses = weighted_losses.view_as(losses)
> >         losses = weighted_losses
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but the loss is very small like 0.0001, I'm not sure if it's normal
> 
> hello, 为啥不直接 flat_losses * torch.exp(-flat_losses).detach()呀，y对应位置的prob已经通过右侧的torch.exp算出来了，直接按位相乘再求和或求均值就行了吧

你说的这样是可以的
